### DISCUSSION TITLE
Idea for configurable hotkeys

### DISCUSSION THREAD

**Discussion by necaran** [tophf/mpiv#119] [MPIV general]
I'm trying to customize hotkeys for my own need.
But currently it's kind of hacky.
Can we make it configurable like this?
```js
let cfg = {
  "keyMapGlobal": {
    "+Shift": "toggleApp"
  },
  "keyMap": {
    "KeyZ": "zoom",
    "KeyX": "fitWidth",
    "KeyC": "fitHeight"
  }
}

const KeyActs = {
  toggleApp() {
  },
  zoom() {
  },
  fitWidth() {
  },
  fitHeight() {
  }
}

const KeyMap = {
  getKeyAct(key) {
    let actName = cfg.keyMapGlobal[key];
    if (!actName && ai.popup)
      actName = cfg.keyMap[key];
    return KeyActs[actName];
  }
}

const Events = {
  onKeyDown(e) {
    if(/INPUT|TEXTAREA|SELECT/.test(e.target.tagName) || e.repeat) return;
    let key = describeKey(e);
    let act = KeyMap.getKeyAct(key);
    if(act instanceof Function && !act(e))
      dropEvent(e);
  }
}
```
By modularizing key mapping from event handling and jsonifying, we can make it somewhat configurable.
Maybe even allow function mode like advanced rules.
A key act can `return true` to allow default event and propagation.

**Comment by tophf** [comment]
Maybe, but to make it useful to more than a couple of users it should provide a convenient way to configure the keys and show the current keys in the help section of the config. I still consider myself a temporary maintainer of the original userscript and I'm fine with the original hotkeys, so I'm not particularly interested. At the moment I think it'd be enough to add a way to configure a custom activation key.
